File: app/src/main/java/com/lifedawn/bestweather/ui/settings/custompreferences/WidgetRefreshIntervalPreference.java
package com.lifedawn.bestweather.ui.settings.custompreferences;

import android.content.Context;
import android.graphics.Color;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.preference.Preference;
import androidx.preference.PreferenceViewHolder;

import com.lifedawn.bestweather.R;

public class WidgetRefreshIntervalPreference extends Preference {
	private TextView textView;
	private Long[] widgetRefreshIntervalLongValues;
	private Long currentValue;
	private int currentValueIndex;
	private String[] widgetRefreshIntervalTexts;

	public WidgetRefreshIntervalPreference(Context context) {
		super(context);
		String[] values = context.getResources().getStringArray(R.array.AutoRefreshIntervalsLong);
		widgetRefreshIntervalLongValues = new Long[values.length];
		int i = 0;

		for (String v : values) {
			widgetRefreshIntervalLongValues[i++] = Long.parseLong(v);
		}
		widgetRefreshIntervalTexts = context.getResources().getStringArray(R.array.AutoRefreshIntervals);
	}

	@Override
	public void onBindViewHolder(PreferenceViewHolder holder) {
		super.onBindViewHolder(holder);
		if (textView == null) {
			textView = new TextView(getContext());
			textView.setText(widgetRefreshIntervalTexts[currentValueIndex]);
			textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 14f);
			textView.setTextColor(Color.BLACK);

			LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT);
			layoutParams.gravity = Gravity.CENTER_VERTICAL;

			textView.setLayoutParams(layoutParams);

			LinearLayout viewGroup = (LinearLayout) holder.findViewById(R.id.layout_widget_root);
			viewGroup.addView(textView);
		} else {
			LinearLayout viewGroup = (LinearLayout) holder.findViewById(R.id.layout_widget_root);
			if (viewGroup.getChildCount() == 0) {
				ViewGroup parent = (ViewGroup) textView.getParent();
				parent.removeView(textView);
				viewGroup.addView(textView);
			}
		}
	}

	public void setValue(Long refreshInterval) {
		this.currentValue = refreshInterval;

		int idx = 0;
		for (Long v : widgetRefreshIntervalLongValues) {
			if (refreshInterval.equals(v)) {
				break;
			}
			idx++;
		}
		currentValueIndex = idx;

		if (textView != null) {
			textView.setText(widgetRefreshIntervalTexts[currentValueIndex]);
		}
	}

	public Long getCurrentValue() {
		return currentValue;
	}

	public Long[] getWidgetRefreshIntervalLongValues() {
		return widgetRefreshIntervalLongValues;
	}

	public String[] getWidgetRefreshIntervalTexts() {
		return widgetRefreshIntervalTexts;
	}

	public int getCurrentValueIndex() {
		return currentValueIndex;
	}
}

File: app/src/main/java/com/lifedawn/bestweather/ui/settings/custompreferences/WidgetRefreshIntervalPreference.kt
package com.lifedawn.bestweather.ui.settings.custompreferences

import android.content.Context
import android.graphics.Color
import android.util.TypedValue
import android.view.Gravity
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import com.lifedawn.bestweather.R

class WidgetRefreshIntervalPreference(context: Context) : Preference(context) {
    private var textView: TextView? = null
    val widgetRefreshIntervalLongValues: Array<Long?>
    var currentValue: Long? = null
        private set
    var currentValueIndex = 0
        private set
    val widgetRefreshIntervalTexts: Array<String>

    init {
        val values = context.resources.getStringArray(R.array.AutoRefreshIntervalsLong)
        widgetRefreshIntervalLongValues = arrayOfNulls(values.size)
        var i = 0
        for (v in values) {
            widgetRefreshIntervalLongValues[i++] = v.toLong()
        }
        widgetRefreshIntervalTexts = context.resources.getStringArray(R.array.AutoRefreshIntervals)
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)
        if (textView == null) {
            textView = TextView(context)
            textView!!.text = widgetRefreshIntervalTexts[currentValueIndex]
            textView!!.setTextSize(TypedValue.COMPLEX_UNIT_SP, 14f)
            textView!!.setTextColor(Color.BLACK)
            val layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT)
            layoutParams.gravity = Gravity.CENTER_VERTICAL
            textView!!.layoutParams = layoutParams
            val viewGroup = holder.findViewById(R.id.layout_widget_root) as LinearLayout
            viewGroup.addView(textView)
        } else {
            val viewGroup = holder.findViewById(R.id.layout_widget_root) as LinearLayout
            if (viewGroup.childCount == 0) {
                val parent = textView!!.parent as ViewGroup
                parent.removeView(textView)
                viewGroup.addView(textView)
            }
        }
    }

    fun setValue(refreshInterval: Long) {
        currentValue = refreshInterval
        var idx = 0
        for (v in widgetRefreshIntervalLongValues) {
            if (refreshInterval == v) {
                break
            }
            idx++
        }
        currentValueIndex = idx
        if (textView != null) {
            textView!!.text = widgetRefreshIntervalTexts[currentValueIndex]
        }
    }
}